spectator-api/src/main/java/com/netflix/spectator/api/patterns/CardinalityLimiters.java

```diff
@@ -193,7 +193,7 @@ private void add(String s) {
         // Lock to keep hashmap consistent with counter for remaining
         lock.lock();
         try {
-          if (remaining.get() > 0) {
+          if (remaining.get() > 0 && !values.containsKey(s)) {
```

Comment:
This can just check if `values.put(s, s) == null`. That way it avoids a second operation on the map in the case it is successful.

Reply:
ah good point, will update.